src/librustc/ich/impls_ty.rs

```diff
@@ -372,10 +375,30 @@ impl_stable_hash_for!(enum ::syntax::ast::Mutability {
     Mutable
 });
 
-impl_stable_hash_for!(struct ty::Const<'tcx> {
-    ty,
-    val
-});
+
+impl<'a, 'gcx> HashStable<StableHashingContext<'a>> for ty::Const<'gcx> {
+    fn hash_stable<W: StableHasherResult>(
+        &self,
+        hcx: &mut StableHashingContext<'a>,
+        hasher: &mut StableHasher<W>,
+    ) {
+        let ty::Const { ty, val, alloc } = self;
+        ty.hash_stable(hcx, hasher);
+        val.hash_stable(hcx, hasher);
+        // don't hash the memory for `Scalar` and `Slice`. There's nothing to be gained
+        // by it. All the relevant info is contained in the value.
+        if let mir::interpret::ConstValue::ByRef = val {
```

Comment:
Uh... so, I know nothing about this stuff.^^ But, isn't it a problem that now we will equate the same constant, when computed multiple times and stored in different allocations? Like, pointer equality might matter for some of them.

Reply:
Well.. the *value* would be the same. Once you have a `ty::Const`, there is no identity anymore. You can't take a reference to a `ty::Const`, so you can't turn it back into a reference.